### DISCUSSION TITLE
Specify minimum libc version

### DISCUSSION THREAD

**Discussion by dr-begemot** [mesonbuild/meson#9726] [Q&A] [answered]
Is it possible to specificy minimum libc version?
For examle like a `dep_libc = dependency('libc', version : '>=2.28')`. 
But problem with libc that it should founded only via `find_library`:
```
cc = meson.get_compiler('c')
dep_libc = dependency('libc', version : '>=2.28', required: false)
if not dep_libc.found()
  dep_libc = cc.find_library('libc', required: true)
endif
```
How I can check libc version?

I try to use [version()](https://mesonbuild.com/Reference-manual_returned_dep.html#depversion)
```
  dep_libc = cc.find_library('libc', required: true)
  if dep_libc.found()  	
  	message ('founded "'+dep_libc.version()+'" libc version')
  endif
```
but it return:
```
meson.build:68:3: ERROR: Unknown method "version" in object <[ExternalLibraryHolder] holds [ExternalLibrary]: <ExternalLibrary libc: True>> of type ExternalLibraryHolder.
```

Thank you!

**Comment by xclaesse** [comment]
Instead of checking for a specific libc version, you probably should check for a specific feature you need. You probably want to check if a function exists, you can do that with cc.has_function().

**Comment by eli-schwartz** [comment, marked as answer]
Dependencies can have version attributes, and check those versions, because pkg-config and cmake etc. have formal methods to expose versions.

find_library cannot and does not have a version attributes, and cannot check versions, because there is no way to extract those versions to begin with.

One possibility would be to add an attribute to check the soname, but this doesn't really tell you anything about the API, only the ABI, so it would be of very little use. Worse, the soname for GNU libc has been "6" for ***decades***. They use symbol versioning rather than library versioning, and this allows preserving ABI compat.

Why do you need to check the GNU libc version? Checking it during configure means that you expect it to be a fatal error if you don't find the "right" version.

Do you not intend to support somewhat older but still actively supported Linux distros on previous versions of the GNU libc? Worse: do you not intend to support other popular libc's like musl at all? What about portability to other operating systems, such as the BSDs?

It seems like it would make far more sense to either check for functions you need, or else do preprocessor checks to see whether GNU libc's version macros are defined and evaluate to the expected version, and then...

... if/else based on that, with some code to take advantage of GNU libc features and some generic code to handle the non-glibc case.

Failing that, you can of course document that only glibc is supported, and do a `cc.compiles()` check that probes the official glibc mechanisms for checking versions.

Either way, please don't manually add `find_library('c')` to your build. That is handled by the compiler.

**Reply by dr-begemot** [reply]
First of all, many thanks for such a detailed answer!!!

In fact I seem to be a little confused.  So what led me to this strange thought to check the libc version. 
I have two OS versions based on buildroot 2018 and 2021, with kernel 3.5 and 5.10 accordingly.
There are deb-package manager (it is actually a modified opkg, but it doesn't matter), any of packages br21 couldn't be installed to br18 due to older libc version. 

This works because the libc version is listed in the deb package and the package simply won't install even if the user tries to do so.

Previously, building a package was carried out in a huge and complex Makefile, which I would like to abandon for a long time.  When I got to know Meson, I found it almost perfect, unlike cmake, and even worse autotools.

In old Makefile I create parts of deb package automatically - `version` and `depends` in control-file, for example.
I was planning to create something similar in Meson - create a special target for generating deb packages, because I don't really like the usual instructions for updating deb packages, there is too much manual work.

I have not yet figured out how much is possible, but for example - go through all the dependencies specified in Meson and add them to the control file.

I hope my explanation was not too messy)
